azurekeyvault-flexvolume/oauth.go

```diff
@@ -191,6 +196,39 @@ func GetServicePrincipalToken(tenantId string, env *azure.Environment, resource
 	return nil, fmt.Errorf("no credentials provided for AAD application %s", aADClientID)
 }
 
+func retryFetchToken(req *http.Request, maxAttempts int) (resp *http.Response, err error) {
+	attempt := 0
+	delay := time.Duration(0)
+	client := &http.Client{}
+	const maxDelay time.Duration = 60 * time.Second
+
+	// retry guidance as suggested here - https://docs.microsoft.com/en-us/azure/active-directory/managed-identities-azure-resources/how-to-use-vm-token#retry-guidance
+	for attempt < maxAttempts {
+		resp, err = client.Do(req)
+
+		if err != nil || resp == nil || resp.StatusCode == http.StatusOK {
```

Comment:
Current retry is for all non successful status codes. We can however streamline this to retry only on specific error codes.